python/phase-1.md

```diff
@@ -0,0 +1,460 @@
+    # specific error to Task/Future.cancel (only in 3.9). So activity
+    # cancellations appear like any other asyncio cancellations (which may be a
+    # good thing anyways).
+    try:
+        while True:
+            await asyncio.sleep(1)
+            activity.heartbeat("some details")
+    except asyncio.CancelledError as exc:
+        raise RuntimeError("someone cancelled me!") from exc
+
+
+class MyCallableClass:
+    def __call__(self):
+        pass
+
+
+class MyActivities:
+    def activity1():
+        pass
+
+
+def create_my_worker(client: temporalio.Client) -> temporalio.Worker:
+    return temporalio.Worker(
+        client,
+        taskqueue="MyTaskQueue",
+        activities={
+            # The names can be any value
+            "simple_activity": simple_activity,
+            # Can use async function
+            "another_activity": another_activity,
+            # Can use callable class
+            "callable_class": MyCallableClass(),
+            # Can use method
+            "activity1_method": MyActivities().activity1,
+            # Can use lambda
+            "lambda": lambda arg1: arg1,
+            # Fails because not callable
+            "not_callable": MyActivities(),
+        },
+    )
+```
+
+Notes:
+
+* Cancellation is task cancellation
+  * We need to research whether, in Python 3.7, we can interrupt an async task w/ our own error reliably
+* Heartbeat is a synchronous call
+* There were a lot of discussions around registering of activities, here are the results:
+  * At least to start with, we want explicit names during register instead of the callable's `__name__` to avoid
+    name-registration surprises that plague other languages
+    * It was decided we can always be more flexible later or provided utilities if we must
+  * A decorator (e.g. `@activity.defn`) was gonna help with naming, but:
+    * Since we decided to use a mapping in the constructor instead of an iterable, name is always there
+    * So the decorator would have no current purpose except as a "marker" to implementers to beware how it is used,
+      which we decided was unnecessary at this time but we can revisit
+  * We still have to have discussions around activity invocation from a worker, because just what appears to be a normal
+    Python invocation turning to an `ExecuteActivity` or not based on decorator or similar alone may not be explicit
+    enough
+
+### Payload Converters
+
+Exported from `temporalio` (but maybe in `temporalio/converter/converter.py`, exported from `temporalio.converter`
+then exported from `temporalio`):
+
+```python
+class PayloadConverter(ABC):
+    # This can be sync or async. Always takes a list. Can return a single
+    # payload or a list of payloads with the same size as the input list or None
+    # if the value cannot be converted with this converter. Any other result
+    # causes an error. Any error raised is bubbled.
+    @abstractmethod
+    async def encode(
+        self, *values: Any
```

Comment:
After thinking about this a bit, we might want to use reflection here to get the expected type from the activity signature like we do in Java / Go

Reply:
I agree. I can mutate this proposal to include that (of course not require type annotations). I will have to do some research on runtime annotation resolution.